### PR TITLE
kluctl: 2.22.1 -> 2.25.0

### PR DESCRIPTION
**Pull request by netthier**
## Description of changes

Changelogs: 
  - https://github.com/kluctl/kluctl/releases/tag/v2.23.0
  - https://github.com/kluctl/kluctl/releases/tag/v2.23.1
  - https://github.com/kluctl/kluctl/releases/tag/v2.23.2
  - https://github.com/kluctl/kluctl/releases/tag/v2.23.3
  - https://github.com/kluctl/kluctl/releases/tag/v2.23.4
  - https://github.com/kluctl/kluctl/releases/tag/v2.23.5
  - https://github.com/kluctl/kluctl/releases/tag/v2.24.0
  - https://github.com/kluctl/kluctl/releases/tag/v2.24.1
  - https://github.com/kluctl/kluctl/releases/tag/v2.25.0

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by netthier** [comment]
Should I add a notice about the breaking changes to the release notes?

**Comment by netthier** [comment]
Did some further testing locally and I'm getting issues involving an unpatched python3 executable, do not merge this.

**Comment by netthier** [comment]
Latest version has made it possible to repack the Python executables and patch them. I've been able to build working versions of kluctl, but am currently trying to fix an issue with the vendorHash not being static.

**Comment by sikmir** [comment]
> Latest version has made it possible to repack the Python executables and patch them.

Good news

**Comment by netthier** [comment]
It seems the reproducibility issues may have been related to my usage of `mktemp` (seems obvious in hindsight...). Removing it seems to result in a consistent vendorHash. I'd appreciate some testing for if this is actually the case and if kluctl works as intended, and I'm also not sure if my `modPostBuild` shenanigans are acceptable, they seem unprecedented in this repo.

**Comment by netthier** [comment]
The plot thickens. I just rebased on latest master, and the `vendorHash` changed. But subsequent builds still show the hash being static. Just as a sanity check, this is what I've tried to confirm that it is static:
1. I ran `nix-collect-garbage -d` with and without sudo
2. I ran `nix-build -A kluctl` and watched the whole build process including the vendor step.
3. I ran `nix-build -A kluctl --check`, which finished successfully.
4. I swapped one character in the `vendorHash` and rebuilt kluctl and its modules again, and the build failed with a mismatched `vendorHash`, showing the original hash that differs by one char as the expected one.
This process showed differences earlier when I still used `mktemp`, but now it does not.

**Comment by netthier** [comment]
Looking at CI again, it does seem like x64 Linux is building successfully! So the hash is platform-specific...
I'll rent out an arm64 VPS to do comparisons next week.

**Comment by netthier** [comment]
https://github.com/kluctl/kluctl/pull/1099 was just merged, meaning I can remove the go-embed-python hack and simplify the build again. I've tested it with kluctl master and it seems to work perfectly.
Once a stable version containing that PR gets released I'll update the commit accordingly and finally move this PR out of draft.

**Comment by netthier** [comment]
The PR is now ready for review/merging; cc @sikmir

File: pkgs/applications/networking/cluster/kluctl/default.nix [hunk 1]
```diff
@@ -1,32 +1,41 @@
...
-  vendorHash = "sha256-EEOVd15f1SK8InSIG+TuVwWibkf+ePJ5AGZpiMD+RaQ=";
+  vendorHash = "sha256-TckT39wQn4dclcYSfxootv1Lw5+iYxY6/wwdUc1+Z6s=";
 
   ldflags = [ "-s" "-w" "-X main.version=v${version}" ];
 
   # Depends on docker
   doCheck = false;
 
+  buildInputs = [
+    pkgs.python3
```
Comment (sikmir) [inline pkgs/applications/networking/cluster/kluctl/default.nix:24]:
```suggestion
    python3
```

File: pkgs/applications/networking/cluster/kluctl/default.nix [hunk 2]
```diff
@@ -1,32 +1,41 @@
-{ lib, stdenv, buildGoModule, fetchFromGitHub, testers, kluctl }:
+{ lib, pkgs, stdenv, buildGoModule, fetchFromGitHub, testers, makeWrapper, kluctl }:
```
Comment (sikmir) [inline pkgs/applications/networking/cluster/kluctl/default.nix:1]:
```suggestion
{ lib, buildGoModule, fetchFromGitHub, testers, makeWrapper, python3, kluctl }:
```

**Comment by sikmir** [comment]
Result of `nixpkgs-review pr 280260` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kluctl</li>
  </ul>
</details>

**Review by MaxwellDupre — APPROVED** [review]
Fairly simple change lgtm.

**Comment by netthier** [comment]
@sikmir Are you able to merge this or is something blocking it?

**Comment by sikmir** [comment]
CI is unhappy on x86_64-darwin as far as I see

**Comment by sikmir** [comment]
@ofborg build kluctl

**Comment by netthier** [comment]
Had to rebase on master due to conflicts

**Comment by sikmir** [comment]
Result of `nixpkgs-review pr 280260` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kluctl</li>
  </ul>
</details>

File: pkgs/applications/networking/cluster/kluctl/default.nix [hunk 3]
```diff
@@ -1,32 +1,41 @@
...
-  vendorHash = "sha256-EEOVd15f1SK8InSIG+TuVwWibkf+ePJ5AGZpiMD+RaQ=";
+  vendorHash = "sha256-TckT39wQn4dclcYSfxootv1Lw5+iYxY6/wwdUc1+Z6s=";
 
   ldflags = [ "-s" "-w" "-X main.version=v${version}" ];
 
   # Depends on docker
   doCheck = false;
 
+  buildInputs = [
+    python3
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/kluctl/default.nix:24]: Why do we need this in buildInputs?

File: pkgs/applications/networking/cluster/kluctl/default.nix [hunk 4]
```diff
@@ -1,32 +1,41 @@
...
+  ];
+
   passthru.tests.version = testers.testVersion {
     package = kluctl;
     version = "v${version}";
   };
 
   postInstall = ''
     mv $out/bin/{cmd,kluctl}
+    wrapProgram $out/bin/kluctl --set KLUCTL_USE_SYSTEM_PYTHON 1 --prefix PATH : '${lib.makeBinPath buildInputs}'
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/kluctl/default.nix:38]:
```suggestion
    wrapProgram $out/bin/kluctl \
      --set KLUCTL_USE_SYSTEM_PYTHON 1 \
      --prefix PATH : '${lib.makeBinPath [ python ]}'
```

File: pkgs/applications/networking/cluster/kluctl/default.nix (hunk 3 above)
Comment (netthier) [reply pkgs/applications/networking/cluster/kluctl/default.nix:24]: I was under the impression that adding it to `buildInputs` was required as it is now a runtime dependency. It seems to still work if I leave out the `buildInputs` and specify `${lib.makeBinPath [ python3 ]}` later.
Can you confirm if that is the change you want?

**Comment by netthier** [comment]
Added the requested changes. I've also pinned Python to 3.10 as this is the version kluctl embeds by default and I noticed some deprecation warnings popping up with 3.11.

File: pkgs/applications/networking/cluster/kluctl/default.nix (hunk 3 above)
Comment (netthier) [reply pkgs/applications/networking/cluster/kluctl/default.nix]: I would be interested in knowing why it _isn't_ needed there though.
Is just mentioning it in `makeBinPath` enough to make it available at runtime?

Comment (SuperSandro2000) [reply pkgs/applications/networking/cluster/kluctl/default.nix]: > Is just mentioning it in makeBinPath enough to make it available at runtime?

yes

**Pull request merged**